protocols/noise/src/io.rs

```diff
@@ -181,32 +190,39 @@ impl<T: io::Read> io::Read for NoiseOutput<T> {
                     trace!("read: copied {}/{} bytes", *off + n, len);
                     *off += n;
                     if len == *off {
-                        self.read_state = ReadState::ReadLen { buf: [0, 0], off: 0 };
+                        this.read_state = ReadState::ReadLen { buf: [0, 0], off: 0 };
                     }
-                    return Ok(n)
+                    return Poll::Ready(Ok(n))
                 }
                 ReadState::Eof(Ok(())) => {
                     trace!("read: eof");
-                    return Ok(0)
+                    return Poll::Ready(Ok(0))
                 }
                 ReadState::Eof(Err(())) => {
                     trace!("read: eof (unexpected)");
-                    return Err(io::ErrorKind::UnexpectedEof.into())
+                    return Poll::Ready(Err(io::ErrorKind::UnexpectedEof.into()))
                 }
-                ReadState::DecErr => return Err(io::ErrorKind::InvalidData.into())
+                ReadState::DecErr => return Poll::Ready(Err(io::ErrorKind::InvalidData.into()))
             }
         }
     }
+
+    unsafe fn initializer(&self) -> futures::io::Initializer {
+        futures::io::Initializer::nop()
+    }
 }
 
-impl<T: io::Write> io::Write for NoiseOutput<T> {
-    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
-        let buffer = self.buffer.borrow_mut();
+impl<T: AsyncWrite + Unpin> AsyncWrite for NoiseOutput<T> {
+    fn poll_write(mut self: std::pin::Pin<&mut Self>, cx: &mut std::task::Context<'_>, buf: &[u8]) -> futures::task::Poll<std::result::Result<usize, std::io::Error>>{
```

Comment:
Why all the namespaces?

```suggestion
    fn poll_write(mut self: Pin<&mut Self>, cx: &mut Context<'_>, buf: &[u8]) -> Poll<Result<usize, std::io::Error>>{
```